client/tests/integration/upgrade.rs

```diff
@@ -195,6 +199,123 @@ fn executor_upgrade_should_revoke_removed_permissions() -> Result<()> {
+
+    // Check that bob has 1 rose
+    assert_eq!(
+        client.request(FindAssetQuantityById::new(bob_rose.clone()))?,
+        Numeric::from(1u32)
+    );
+
+    // Give 1 rose to all
+    let isi = MintAssetForAllAccounts {
+        asset_definition_id,
+        quantity: Numeric::from(1u32),
+    };
+    let isi = CustomInstructionBox::MintAssetForAllAccounts(isi);
+    let isi = Custom::new(serde_json::to_string(&isi)?);
+    client.submit_blocking(isi)?;
+
+    // Check that bob has 2 roses
+    assert_eq!(
+        client.request(FindAssetQuantityById::new(bob_rose.clone()))?,
+        Numeric::from(2u32)
+    );
+
+    Ok(())
+}
+
+#[test]
+fn executor_custom_isi_complex() -> Result<()> {
+    use executor_custom_data_model::complex::{
+        ConditionalExpr, CoreExpr, CustomInstructionExpr, EvaluatesTo, Expression, Greater,
+    };
+    use iroha_config::parameters::actual::Root as Config;
+
+    let mut config = Config::test();
+    // Note that this value will be overwritten by genesis block with NewParameter ISI
+    // But it will be needed after NewParameter removal in #4597
+    config.chain_wide.executor_runtime.fuel_limit = 1_000_000_000;
+
+    let (_rt, _peer, client) = PeerBuilder::new()
+        .with_port(11_185)
+        .with_config(config)
+        .start_with_runtime();
+    wait_for_genesis_committed(&vec![client.clone()], 0);
+
+    // Remove this after #4597 - config value will be used (see above)
+    let parameters = ParametersBuilder::new()
+        .add_parameter(EXECUTOR_FUEL_LIMIT, Numeric::from(1_000_000_000_u32))?
+        .into_set_parameters();
+    client.submit_all_blocking(parameters)?;
+
+    upgrade_executor(
+        &client,
+        "tests/integration/smartcontracts/executor_custom_isi_complex",
+    )?;
+
+    // Give 6 roses to bob
+    let asset_definition_id: AssetDefinitionId = "rose#wonderland".parse().unwrap();
+    let bob_rose = AssetId::new(asset_definition_id.clone(), BOB_ID.clone());
+    client.submit_blocking(Mint::asset_numeric(Numeric::from(6u32), bob_rose.clone()))?;
+
+    // Check that bob has 6 roses
+    assert_eq!(
+        client.request(FindAssetQuantityById::new(bob_rose.clone()))?,
+        Numeric::from(6u32)
+    );
+
+    // If bob has more then 5 roses, then burn 1 rose
+    let burn_bob_rose_if_more_then_5 = || -> Result<()> {
+        let condition = Greater::new(
+            EvaluatesTo::new_unchecked(Expression::Query(
+                FindAssetQuantityById::new(bob_rose.clone()).into(),
+            )),
+            Numeric::from(5u32),
+        );
+        let then = Burn::asset_numeric(Numeric::from(1u32), bob_rose.clone());
+        let then: InstructionBox = then.into();
+        let then = CustomInstructionExpr::Core(CoreExpr::new(then));
+        let isi = CustomInstructionExpr::If(Box::new(ConditionalExpr::new(condition, then)));
+        let isi = Custom::new(serde_json::to_string(&isi)?);
+        client.submit_blocking(isi)?;
```

Comment:
```suggestion
        client.submit_blocking(isi)?;
```

is it possible to implement `From<CustomInstructionExpr> for InstructionBox`? That would reduce friction in the API and make the conversion to json implicit

Reply:
also add impl `From<CustomInstructionExpr> for Custom`

Reply:
To use `client.submit()` for `CustomInstructionExpr`, it is necessary for `CustomInstructionExpr` to implement `Instruction` and `Sealed` traits. This will require some changes to iroha_data_model (at least make `Sealed` trait public). I am not sure if it is worth it.

Currently I added `pub fn into_instruction(self) -> InstructionBox` method to `CustomInstructionExpr`, so it is possible to write:
```rust
let isi = CustomInstructionExpr::If(...);
client.submit_blocking(isi.into_instruction())?;
```

Reply:
I created #4703